include/cuco/detail/trie/bit_vector/bit_vector.cuh

```diff
@@ -0,0 +1,200 @@
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuco/detail/trie/bit_vector/bit_vector_ref.cuh>
+#include <cuco/extent.cuh>
+#include <cuco/storage.cuh>
+#include <cuco/utility/allocator.hpp>
+
+#include <thrust/device_vector.h>
+
+#include <cuda/atomic>
+
+namespace cuco {
+namespace experimental {
+
+/**
+ * @brief Struct to store ranks of bits at 256-bit intervals
+ */
+struct rank {
+  uint32_t abs_hi_;              ///< Upper 32 bits of base
+  uint8_t abs_lo_;               ///< Lower 8 bits of base
+  std::array<uint8_t, 3> rels_;  ///< Offsets for 64-bit sub-intervals
+
+  /**
+   * @brief Gets base rank of current 256-bit interval
+   *
+   * @return The base rank
+   */
+  __host__ __device__ uint64_t abs() const
+  {
+    return (static_cast<uint64_t>(abs_hi_) << 8) | abs_lo_;
+  }
+
+  /**
+   * @brief Sets base rank of current 256-bit interval
+   *
+   * @param abs Base rank
+   */
+  void set_abs(uint64_t abs)
+  {
+    abs_hi_ = static_cast<uint32_t>(abs >> 8);
+    abs_lo_ = static_cast<uint8_t>(abs);
+  }
+};
+
+/**
+ * @brief Union of 64-bit word with rank
+ *
+ * Need this so that all aow_storage structures in bitvector have 64-bit element type
+ */
+union rank_union {
+  uint64_t word_;  ///< word view
+  rank rank_;      ///< rank view
+};
+
+/**
+ * @brief Bitvector class with rank and select index structures
+ *
+ * In addition to standard bitvector get/set operations, this class provides
+ * rank and select operation API. It maintains index structures to make both these
+ * new operations close to constant time.
+ * Bitvector construction happens on host, after which the structures are moved to device.
+ * All subsequent read-only operations access device structures only.
+ *
+ * @tparam Extent Data structure size type
+ * @tparam Scope The scope in which operations will be performed by individual threads.
+ * @tparam Allocator Type of allocator used for device storage
+ * @tparam Storage Slot window storage type
+ */
+
+template <class Extent             = cuco::experimental::extent<std::size_t>,
+          cuda::thread_scope Scope = cuda::thread_scope_device,
+          class Allocator          = cuco::cuda_allocator<std::byte>,
+          class Storage            = cuco::experimental::storage<1>>
+class bit_vector {
+ public:
+  bit_vector();
```

Comment:
```suggestion
  bit_vector(Allocator const& allocator = Allocator{});
```
We need to at least pass a default allocator

Reply:
a371e27b6c612e7a61bee2fe6a6e70654f006051